internal/otlptext/databuffer.go

```diff
@@ -128,6 +133,54 @@ func (b *dataBuffer) logDoubleHistogramDataPoints(ps pdata.HistogramDataPointSli
 	}
 }
 
+func (b *dataBuffer) logExponentialHistogramDataPoints(ps pdata.ExponentialHistogramDataPointSlice) {
+	for i := 0; i < ps.Len(); i++ {
+		p := ps.At(i)
+		b.logEntry("ExponentialHistogramDataPoints #%d", i)
+		b.logDataPointAttributes(p.Attributes())
+
+		b.logEntry("StartTimestamp: %s", p.StartTimestamp())
+		b.logEntry("Timestamp: %s", p.Timestamp())
+		b.logEntry("Count: %d", p.Count())
+		b.logEntry("Sum: %f", p.Sum())
+
+		scale := int(p.Scale())
+		factor := math.Ldexp(math.Ln2, -scale)
+		// Note: the equation used here, which is
+		//   math.Exp(index * factor)
+		// reports +Inf as the _lower_ boundary of the bucket nearest
+		// infinity, which is incorrect and can be addressed in various
+		// ways.  The OTel-Go implementation of this histogram pending
+		// in https://github.com/open-telemetry/opentelemetry-go/pull/2393
+		// uses a lookup table for the last finite boundary, which can be
+		// easily computed using `math/big` (for scales up to 20).
+
+		negB := p.Negative().BucketCounts()
+		posB := p.Positive().BucketCounts()
+
+		for i := 0; i < len(negB); i++ {
+			pos := len(negB) - i - 1
+			index := p.Negative().Offset() + int32(pos)
+			count := p.Negative().BucketCounts()[pos]
+			lower := math.Exp(float64(index) * factor)
+			upper := math.Exp(float64(index+1) * factor)
```

Comment:
These calculations are inherent property of how we define the histogram, right? Ideally we would want the calculation logic to be encapsulated in the `ExponentialHistogramDataPoint` struct and be callable as methods here. I am not sure if it is easily doable since the struct I think is auto-generated. As an alternate perhaps extract this logic as common helper functions somewhere so that any code that needs to read the histogram data can reuse it.

Reply:
I was a bit conflicted about this -- the code in open-telemetry/opentelemetry-go#2393 has these equations written in an internal package, and if those packages were exposed we wouldn't need to implement them here.

There is hand-written code in the `model/pdata` package, so we can add methods outside of the generated code.

OTOH, I had been thinking of this as a low-level dependency, and a non-critical use of the exponential histogram where it does not warrant bringing in a larger library. I documented how this equation is not perfect near infinity, which I think is a detail that does not matter for a logging exporter. I'd also be happy to create helper methods in  the `model/pdata.ExponentialHistogramDataPoint` using these equations for now, leave a TODO stating to replace the hard-coded equation after open-telemetry/opentelemetry-go#2393 or an equivalent library merges.

Reply:
I will follow up.